[chua wen hong] ip

runoutofit opened the pull request:
# Duke
Duke frees your mind by remembering things you need to do
 
 It's
 
 * Text-based
 * Easy to learn
 * ~Fast~ _SUPER FAST_ to use
 
 ## Steps to run
 1. Pull the code from the repo [here](https://github.com/runoutofit/ip)
 2. Run it
 3. Add your tasks
 4. Let it manage your tasks for you 
 
 ## Features
 * [x]  Managing todo
 * [x]  Managing deadlines
 * [x]  Managing events
 * [x]  Mark and unmark tasks
 
 This is the main function:
 
 ```ruby
 public static void main(String[] args) throws DukeException, IOException {
        new Duke("data/duke.txt").run();
    }
 ```

Commit:
Add Gradle support

Commit:
level 1

Commit:
level 2 add and list

Commit:
level 3 Mark as Done

Commit:
level 4 todos,deadlines and events

Commit:
automated testing

Commit:
automated testing and fixed Duke

Commit:
Merge remote-tracking branch 'origin/master'

Commit:
add Level 5

Commit:
add level 6-delete

jamieeeleow commented on src/main/java/Duke.java:
```diff
@@ -1,4 +1,6 @@
 import java.util.Scanner;
+import java.util.ArrayList;
+import java.util.List;
```
I like that all the imported classes are listed explicitly. Perhaps you could consider ordering them in alphabetical order?
```ruby
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;
```

jamieeeleow commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,26 @@
+
+    public void markAsDone() {
+        this.isDone = true;
+    }
```
I like how neat this part of the code is. I think the naming of the methods makes it clear what the function does.

Commit:
add level 7-save

jamieeeleow commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,14 @@
+public class Deadline extends Task {
+
+    protected String by;
```
Perhaps the modifier for the variable ```by``` could be ```private```?

jamieeeleow commented on src/main/java/Duke.java:
```diff
@@ -6,70 +6,81 @@ public class Duke {
                 }
                 echo = sc.nextLine();
-            } else if ((echo.length() > 4) && (echo.substring(0,4).equals("mark"))) {
+            } else if (echo.equals("bye")) {
```
I like the change of code in identifying the type of command, makes the entire code clean and neat.

jamieeeleow commented on src/main/java/UnknownCommandException.java:
```diff
@@ -0,0 +1,11 @@
+        super("OOPS!!! I'm sorry, but I don't know what that means :-(\n");
+    }
+
+    @Override
+    public String toString() {
```
Perhaps this ```toString()``` method could be removed for the exception classes? The message could be retrieved with the built-in method of the Exception class buy using ```getMessage()``` function.

jamieeeleow commented on src/main/java/Duke.java:
```diff
@@ -1,27 +1,86 @@
 public class Duke {
+    private static List<Task> store = new ArrayList<>();
 
+    private static void readSaveFile(String filePath) throws FileNotFoundException, EmptyDescriptionException {
```
I like that the methods are private since they are only used within the class itself 😄👍

Commit:
add level 8-date and time

Commit:
Merge branch 'branch-level-8'

# Conflicts:
#	src/main/java/Duke.java

Commit:
fix merging issues

Commit:
Add more OOP

Commit:
Add duke package

Jayanth-Balasubramanian commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,43 @@
+        String cmd = ui.readCommand();
+        boolean isTerminated = false;
+        while (!isTerminated) {
+            if (!cmd.equals("bye")) {
```
Consider making the happy path prominent. 
if (cmd.equals("bye")) {
    // your cleanup code
    return
}
// main path here

Jayanth-Balasubramanian commented on src/main/java/Parser.java:
```diff
@@ -0,0 +1,54 @@
+public class Parser {
+    private static Ui ui = new Ui();
+    public static void parse(String cmd, TaskList list) throws DukeException {
```
Nitpicking, but this method is a little long, you might want to split it into multiple ones such as handleList, handleMark etc so it's easier to debug

Jayanth-Balasubramanian commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,31 @@
+    protected String description;
+    protected boolean isDone;
+
+    public Task(String description) throws EmptyDescriptionException{
```
I like how specific the exception is, definitely helps the user!

Jayanth-Balasubramanian commented on src/main/java/Ui.java:
```diff
@@ -0,0 +1,59 @@
+import java.util.Scanner;
+
+public class Ui {
+    private static final String welcomeMsg = "Hello Boss.\n" + "How may i help you?\n";
```
this looks like a constant, it should be in SCREAMING_SNAKE_CASE

Jayanth-Balasubramanian commented on src/main/java/Ui.java:
```diff
@@ -0,0 +1,59 @@
+
+public class Ui {
+    private static final String welcomeMsg = "Hello Boss.\n" + "How may i help you?\n";
+    private static final String line = "__________________________________________\n";
```
This should also be in screaming snake case as above

Jayanth-Balasubramanian commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,36 @@
+
+    public Deadline(String description, String by) throws EmptyDescriptionException {
+        super(description);
+        String[] str = by.split(" ");
```
Variable name could be more specific. Maybe something like String[] tokens = by.split(" ")?

Jayanth-Balasubramanian reviewed (commented):
LGTM, apart from a few minor nitpicks. Very clean code!

Commit:
Merge remote-tracking branch 'origin/add-gradle-support'

Commit:
add JUnit tests-test deadline and toDo

bryanongjx commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,36 @@
+import java.time.LocalTime;
+import java.time.format.DateTimeFormatter;
+
+public class Deadline extends Task {
```
can consider adding documentation to your classes for readers to understand

bryanongjx commented on src/main/java/duke/DukeException.java:
```diff
@@ -0,0 +1,12 @@
+package duke;
```
you should leave a space between your imports and the class

bryanongjx commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,32 @@
+    protected boolean isDone;
+
+    public Task(String description) throws EmptyDescriptionException{
+        if (description.equals("") || description.equals(" ")) {
```
can consider using .isEmptyString instead of .equals("")

bryanongjx commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,42 @@
+        }
+    }
+
+    public static ArrayList<Task> load() throws DukeException, FileNotFoundException {
```
good that you separated the 2 exceptions! 👍

Commit:
add javadocs

Commit:
add coding standards

Commit:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/Parser.java

Commit:
add find feature

Commit:
commit duke txt

Commit:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/Duke.java
#	src/main/java/duke/Parser.java

Commit:
resolve merge conflict

Commit:
fix unnoticed error with storage

Commit:
fix and tidy up code

Commit:
resolve warnings

Commit:
add Level 10

Commit:
merge Level 10

Commit:
clean  up some code and add some javadocs

Commit:
add assert

Commit:
add codeQuality

runoutofit committed:
Merge pull request #2 from runoutofit/branch-A-Assertions

add assert

runoutofit committed:
Merge pull request #3 from runoutofit/branch-A-CodeQuality

add codeQuality

Commit:
add BCDextension C-FriendlierSyntax

previous version requires user to type full command.

friendlier syntax is implemented for detecting input commands.

let's include use of one letter commands to be detected as INPUT.

Commit:
clean some code

Commit:
update gui

Commit:
resolve some obvious warnings

Commit:
add README

Commit:
add minor changes

Commit:
correct filepath

Commit:
adjust dialog box

Commit:
adjust main window

Commit:
add Ui.png and update deadline error message

Commit:
add Duke.png

Commit:
update javaDocs

Commit:
update javaDocs, abstract code to ui, other minor changes
